src/gt4py/next/iterator/transforms/inline_lifts.py

```diff
@@ -139,125 +164,208 @@ class Flag(enum.IntEnum):
-            is_lift = _is_lift(node.args[0])
-            is_eligible = is_lift and self.predicate(node.args[0], is_scan_pass_context)
-            is_trivial = is_lift and len(node.args[0].args) == 0  # type: ignore[attr-defined] # mypy not smart enough
-            if (
-                self.flags & self.Flag.INLINE_DEREF_LIFT
-                or (self.flags & self.Flag.INLINE_TRIVIAL_DEREF_LIFT and is_trivial)
-            ) and is_eligible:
-                assert isinstance(node.args[0], ir.FunCall) and isinstance(
-                    node.args[0].fun, ir.FunCall
-                )
-                assert len(node.args[0].fun.args) == 1
-                f = node.args[0].fun.args[0]
-                args = node.args[0].args
-                new_node = ir.FunCall(fun=f, args=args)
-                if isinstance(f, ir.Lambda):
-                    new_node = inline_lambda(new_node, opcount_preserving=True)
-                return self.visit(new_node, **kwargs)
-        elif self.flags & self.Flag.PROPAGATE_CAN_DEREF and node.fun == ir.SymRef(id="can_deref"):
-            # TODO(havogt): this `can_deref` transformation doesn't look into lifted functions,
-            #  this need to be changed to be 100% compliant
-            assert len(node.args) == 1
-            if _is_lift(node.args[0]) and self.predicate(node.args[0], is_scan_pass_context):
-                assert isinstance(node.args[0], ir.FunCall) and isinstance(
-                    node.args[0].fun, ir.FunCall
+    def transform_propagate_shift(self, node: ir.FunCall, *, recurse, **kwargs):
+        if not _is_shift_lift(node):
+            return None
+        shift = node.fun
+        assert len(node.args) == 1
+        lift_call = node.args[0]
+        new_args = [
+            self.visit(ir.FunCall(fun=shift, args=[arg]), recurse=False, **kwargs)
+            for arg in lift_call.args  # type: ignore[attr-defined] # lift_call already asserted to be of type ir.FunCall
+        ]
+        result = ir.FunCall(fun=lift_call.fun, args=new_args)  # type: ignore[attr-defined] # lift_call already asserted to be of type ir.FunCall
+        # TODO: describe everywhere
+        copy_recorded_shifts(from_=node, to=result, required=False)
+        return self.visit(result, recurse=False, **kwargs)
+
+    def transform_inline_deref_lift_impl(self, node: ir.FunCall, **kwargs):
+        # TODO: this does not work for neighbors(..., lift(...)), which is essentially also a deref
+        assert isinstance(node.args[0], ir.FunCall) and isinstance(node.args[0].fun, ir.FunCall)
+        assert len(node.args[0].fun.args) == 1
+        f = node.args[0].fun.args[0]
+        args = node.args[0].args
+        new_node = ir.FunCall(fun=f, args=args)
+        if isinstance(f, ir.Lambda):
+            new_node = inline_lambda(new_node, opcount_preserving=True)
+        return self.visit(new_node, **kwargs)
+
+    def transform_inline_deref_lift(self, node: ir.FunCall, **kwargs):
```

Comment:
please describe what `None` result means

Reply:
I added a comment in `visit_FunCall` describing the meaning. Let me know if you want more.